Refactor/alper keskin/clean code

alprKeskin opened the pull request:
Refactoring for clean code

alprKeskin committed:
Clean Code Refactor

Clean Code Refactor

* Class renaming
* Class segregation

alprKeskin committed:
Revert "Clean Code Refactor"

This reverts commit d12c7e28008ac66e842cb747dbd0f8b41fc8d78b.

alprKeskin committed:
Revert "Revert "Clean Code Refactor""

This reverts commit 9b16413eb0d91b3fff1bc20d0e49acd23d91d927.

alprKeskin committed:
Clean Code Refactor

*Class Renaming
*Class Segregation

alprKeskin committed:
Removed duplications

- Runtime exception has been added for a future date input
- Reduced the endpoint number

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/OpenExchangeRatesMicroserviceApplication.java:
```diff
@@ -1,13 +1,40 @@
+// Bunu ve getDailyRates'i ayrı bir class'ta define etme şansımız var mı?
+@Configuration
+@EnableScheduling
+@ConditionalOnProperty(name = "scheduling.enabled", matchIfMissing = true)
```
nice :)

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/OpenExchangeRatesMicroserviceApplication.java:
```diff
@@ -1,13 +1,40 @@
+    // second minute hour day_of_month month day_of_week
+    @Scheduled(cron = "*/10 * * * * *")
+    public void getDailyRates() {
+        System.out.println("Scheduled getDailyRates...");
```
println yerine classlarına bir adet static Logger LOGGER = LoggerFactory.getLogger(Class'ın); şeklinde tanımlayıp, ilgili yerlerde de LOGGER.info(...) LOGGER.error(...) gibi kullanırsan daha profesyonel olur.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/OpenExchangeRatesMicroserviceApplication.java:
```diff
@@ -1,13 +1,40 @@
         SpringApplication.run(OpenExchangeRatesMicroserviceApplication.class, args);
     }
 
+    // OpenExchangeRatesService::getResponse()'u static mi yapsak?
```
Bean'ler zaten singelton hareket ettiği için static yapmaya pek gerek yok. Bean yapısı içerisinde static methodlar pek de fine kabul edilmemekte. Static bir method kullanacaksan class'ının final class olarak tanımlanmasında fayda var.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -1,40 +1,43 @@
-import static org.springframework.http.ResponseEntity.ok;
+import java.time.LocalDate;
 
+// Bir http isteği geldiğinde bu class'ın method'ları otomatik olarak path'e göre çalışacaktır.
```
ingilizce comment bırakırsan alışkanlık olarak daha iyi olur. bir de bu tarz açıklayıcı infolar yazacaksan bilgilendirme commenti syntax'ı kullanırsan daha iyi olur. bu nedir dersen commenti // bunun ile başlatmak yerine /** comment **/ şeklinde yaparsan olacaktır. Yalnız tek yıldız değil çift yıldız.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -1,40 +1,43 @@
-        else {
-            return ok(service.getAndSaveLatest(apiEndpoint, app_id, base, symbols, prettyprint, show_alternative));
-        }
+    MainService mainService;
```
private olmalı

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -1,40 +1,43 @@
+
+    // latest endpoint
+    @GetMapping("/{latestEndpoint}")
+    public ResponseEntity<CurrencyRates> getResponseLatest(
```
burada isteğim aslında şu 
@GetMapping("/{date}")
public ResponseEntity<CurrencyRates> getCurrencies(@PathVariable(value = "date", required = false), String dateStr,
            @RequestParam(value = "symbols", required = false, defaultValue = "XXX") String symbols) {
        LocalDate date =  dateStr == null || dateStr.length != 10 ? LocalDate.now() : 
                          LocalDate.parse(historicalEndpoint.substring(0,10));

        return mainService.getResponse(date, symbols);       
}

dersen iki endpoint değil tek endpointte yazmış olursun. ayrıca service class'ının adını daha anlaşılır yapsan faydalı olur. bir de senin get response'da tek ihtiyacın olan date ve symbols üçüncü bir parametre lazım değil senin için. Eğer onu db kaydı için yolluyorsan gerek yok LocalDate'ten istediğin formatta çekebilirsin string'i
            
)

alprKeskin replied:
Bu feedback'i göremedim ilk attığında abi. Şu anda bununla ilgili bir problem var mı? Bu dediğin gibi şu an. Sadece if else haliyle yazılı. Bir de şu var @GetMapping("/{date}") şeklinde map'lediğimde tarih verilmiş url'i kabul ediyor. Ancak şunları kabul etmiyor.
=> https://localhost8082/api
=> https://localhost8082/api/
Yani tarih verilmediği durumu kabul etmiyor. (Not found diyor.)
Bu sebeple şöyle yaptım. @GetMapping(value = {"", "/{date}"})
Bu istediğin gibi mi?

berkayelken replied:
yes, bu da olur ya da {} içine default değer koyma gibi şeyler de var ama bu daha doğru bu case için

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/DatabaseService.java:
```diff
@@ -0,0 +1,33 @@
+// @Service annotation creates a bean of the service class
+// So you can use @Autowired
+@Service
+public class DatabaseService {
```
buranın adını CurrencyPersistanceService gibi bir şey yaparsan daha hoş olur

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/ExchangeRatesService.java:
```diff
@@ -28,47 +28,33 @@ public class ExchangeRatesService {
         String app_idPart = "app_id=" + app_id;
-        String basePart = "&base=" + base;
         String symbolsPart;
         if (symbols.equals("XXX")) {
```
equals controlü constant olandan başlanırsa daha robust olur her zaman için yani örneğin symbols null geldi şu an null pointer yersin ama eğer kontrolü "XXX".equals(symbols) dersen sorun ortadan kalkar

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/MainService.java:
```diff
@@ -0,0 +1,42 @@
+import static org.springframework.http.ResponseEntity.ok;
+
+@Service
+public class MainService {
```
would be renamed

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/MainService.java:
```diff
@@ -0,0 +1,42 @@
+    private DatabaseService databaseService;
+
+
+    public ResponseEntity<CurrencyRates> getResponse(String endpoint, String symbols, LocalDate date) {
```
parametreler için LocalDate date, String Symbols yeterli.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/MainService.java:
```diff
@@ -0,0 +1,42 @@
+        // try to take the currency rates of the related date from database
+        Optional<CurrencyRates> currencyRates = databaseService.findCurrencyRatesById(date);
+        // if there is no any current rates object of the related date in the database
+        if (currencyRates.equals(Optional.empty())) {
```
bunun yerine !currencyRates.isPresent() kullanmalısın.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/MainService.java:
```diff
@@ -0,0 +1,42 @@
+            return ok(currencyRatesOfRelatedDate);
+        }
+        // if there exist a current rates object of related date in the database
+        else {
```
if'te return ettiğin için if'e girerse zaten kod'a devam etmeyecek. dolayısıyla else kısmını yazmak lazım değil. else içine yazdığını direkt if'in devamına bir scope'a almadan yazabilirsin.

alprKeskin committed:
İmprovements

- Logger usage instead of println method
- rewrited comments correctly
- equal check has been improved
- Main Service has been renamed
- Unnecessary parameters has been removed

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/OpenExchangeRatesMicroserviceApplication.java:
```diff
@@ -18,21 +20,24 @@ public static void main(String[] args) {
 
-    // OpenExchangeRatesService::getResponse()'u static mi yapsak?
+
+    private Logger logger = Logger.getLogger(OpenExchangeRatesMicroserviceApplication.class.getName());
```
sadece .class'a kadar yazman yeterli

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -1,41 +1,52 @@
-    MainService mainService;
+    private OpenExchangeRatesMainService openExchangeRatesMainService;
+
+    Logger logger = Logger.getLogger(ServiceController.class.getName());
```
ServiceController.class yeterli

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -1,41 +1,52 @@
-            @RequestParam(value = "symbols", required = false, defaultValue = "XXX") String symbols){
-        // modify the desired day properly
-        LocalDate desiredDate = LocalDate.parse(historicalEndpoint.substring(0,10));
+        LocalDate desiredDate;
```
bu kısımdaki logic'i ayrı bir method'a ayrı bir class'ta çıksan daha iyi olur. Örneğin DateUtil diye bir final class oluştur onun içine String'ten LocalDate veren bir static method yaz, böylece daha okunur ve anlaşılır bir kodun olur.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -1,41 +1,52 @@
+            // modify the requested date properly
+            desiredDate = LocalDate.parse(requestedDate.substring(0,10));
+            // if the desired date is in the future of the current date
+            if (desiredDate.compareTo(LocalDate.now()) >= 0) {
```
compare to tam istediğini veriyor mu emin değilim ama araştırman sonucu böyle bir bilgiye ulaştıysan kalabilir.

alprKeskin committed:
improvements_2

- added DateUtils
- cleaner code in the service controller

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/rest/ServiceController.java:
```diff
@@ -17,35 +18,14 @@ public class ServiceController {
     private OpenExchangeRatesMainService openExchangeRatesMainService;
 
-    Logger logger = Logger.getLogger(ServiceController.class.getName());
+    Logger logger = Logger.getLogger("ServiceController.class");
```
String literal kullanmana gerek yok şu şekilde tanımlaman daha doğru olur:

private static final Logger LOGGER = Logger.getLogger(ServiceController.class);

alprKeskin replied:
Abi dediğin gibi yazdığımda hata veriyor. Logger.getLogger parametre olarak string alır diyor.

berkayelken commented on src/main/java/io/github/alprkeskin/openexchangeratesmicroservice/service/OpenExchangeRatesMainService.java:
```diff
@@ -16,19 +16,19 @@ public class OpenExchangeRatesMainService {
-        Optional<CurrencyRates> currencyRates = databaseService.findCurrencyRatesById(date);
+        Optional<CurrencyRates> currencyRates = currencyPersistanceService.findCurrencyRatesById(date);
         // if there is no any current rates object of the related date in the database
         if (!currencyRates.isPresent()) {
```
aslında isPresent'ı ben dedim ama direkt şöyle de kullanabilirsin:

if (currencyRates.isEmpty())

alprKeskin replied:
isEmpty ile değiştirdim. Daha düzgün oldu.

alprKeskin committed:
Logger and optional methods

- Made logger private final static
- Used isEmpty() instead of !isPresent()

alprKeskin committed:
Logger

- used slf4j logger instead of java.util.Logger
